src/taskgraph/util/vcs.py

```diff
@@ -511,8 +511,11 @@ def get_repository(path):
             return HgRepository(path)
         elif os.path.exists(os.path.join(path, ".git")):
             return GitRepository(path)
-
-    raise RuntimeError("Current directory is neither a git or hg repository")
+        elif os.path.exists(os.path.join(path, "moz.configure")):
+            return None
+    raise RuntimeError(
+        "Current directory is neither a git or hg repository, nor a release source"
+    )
```

Comment:
I don't love putting Gecko specific logic here. Do you know where `get_repository` was being called from? We've established that the invocation in `parameters.py` can be bypassed. That only leaves `files_changed.py` (which definitely needs a VCS to work anyway), `decision.py` (which I wouldn't suspect would be used for your use case) and `main.py`.

Perhaps we could instead update `main.py` to catch `RuntimeError`? Maybe we could invent a new `Exception` class to better handle it too.

Reply:
Or failing that, I'd rather we always return `None` here, and then assert that `repo is not None` for any existing callers that are expecting it.

Reply:
(although that latter proposal is backwards incompatible, so would require a major version bump.. so I'm leaning towards the former)

Reply:
`get_repository()` is called from `_get_defaults()`

If i run `./mach bootstrap` without any of the changes to taskgraph, what happens is that, the bootstrap passes through, installing os dependencies and python dependencies but the artifacts like `clang`,`clang-tidy`,`nasm` etc. is not installed, which could cause issues with `./mach build`:ing

for those to install i need to have the changes in `vcs.py` and `parameters.py`

I will dig deeper. i see where you come from, touch as little as possible in `taskgraph` :)
which i understand now, when this is not part of mozilla-central, but its own entity (which i didn't realize earlier until a few weeks ago :))

I have to remember how things held together, have forgotten that when the patch was pushed.
I get back here when i know more and have some better answers.

Reply:
hey @ahal 
I might have an idea. If it sounds good, we close this PR and i create a follow up bug in mozilla-bugzilla. mention this pr and cc you.

This is the callstack prettymuch for this.

maybe_install_private_packages_or_exit() is called from bootstrap() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozboot/mozboot/bootstrap.py#348
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozboot/mozboot/bootstrap.py#394

install_toolchain_artifact() is called from maybe_install_private_packages_or_exit()  in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozboot/mozboot/bootstrap.py#263

bootstrap_toolchain() is called from install_toolchain_artifact()  in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozboot/mozboot/base.py#347

bootstrap_path() is called from bootstrap_toolchain() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozbuild/mozbuild/bootstrap.py#52

bootstrap_toolchain_tasks() is called from bootstrap_path() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/build/moz.configure/bootstrap.configure#151

toolchain_task_definitions() is called from bootstrap_toolchain_tasks()  in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/build/moz.configure/bootstrap.configure#110

load_tasks_for_kind() is called from toolchain_task_definitions() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozbuild/mozbuild/toolchains.py#20

parameters_loader() is called from load_tasks_for_kind() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/generator.py#444

load_parameters_file() is called from parameters_loader() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#359

Parameters.__init__() is called from load_parameters_file() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#310

fill_defaults() is called from Parameters.__init__() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#159

get_defaults() via defaults_functions is called from fill_defaults() or extend_parameters_schema in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#146
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#198

get_repository() is called from get_defaults() in:
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/parameters.py#82

get_repository()
https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/third_party/python/taskcluster_taskgraph/taskgraph/util/vcs.py#505


And im thinking i would, just above https://searchfox.org/mozilla-central/rev/ddff4872b8220fae71155e39fd30ac3b760a4741/python/mozbuild/mozbuild/toolchains.py#20
call ´get_repository()` in a `try` / `except RuntimeError`.
If it hits `RuntimeError` i would check if a `moz.configure` file exists. Then make whatever calls i need from above callstack from the taskcluster things, either by hand or calling the actual functions from `python/mozbuild/mozbuild/toolchains.py`

If we dont hit the exception, we continue as usual.
Meaning no touching the `third_party/python/taskcluster_taskgraph/` stuff.

Makes sense?

Reply:
Yes, that makes sense to me! And I think it's the right approach. While I'm not opposed to supporting running Taskgraph without a source checkout, it is very much an edge case and adds a fair amount of complexity. E.g, do only the paths touched by `./mach bootstrap` need to support it? Or do all transforms need to support it (I'd guess there are many places that assume we have a VCS in those)?

Detecting this in the code that consumes Taskgraph and taking appropriate action sounds great! If you're hitting errors and need some helping figuring out how to workaround them, let me know! (Though please use needinfo or ping me or I may miss it :))